### ISSUE TITLE
Same seed and dimlets result in 2 different dimensions

### ISSUE THREAD

**Issue by josephcsible** [McJty/RFToolsDimensions#155]
I started Minecraft via `./gradlew runClient` with git version f0525e9. I created a new world that happened to have seed `-9201105177961642346`. I created a dimension with no dimlets, which spawned me above-ground in a flower forest with a bunch of pillars. Info:
```
[13:26:50] [main/INFO]: [CHAT] §e    Base seed: -9201105177961642346
[13:26:50] [main/INFO]: [CHAT] §e    World version: 1
[13:26:50] [main/INFO]: [CHAT] §e    Terrain: TERRAIN_ROUGH
[13:26:50] [main/INFO]: [CHAT] §e        Base block: Stone
[13:26:50] [main/INFO]: [CHAT] §e        Canyon block: Stone
[13:26:50] [main/INFO]: [CHAT] §e        Base fluid: Air
[13:26:50] [main/INFO]: [CHAT] §e    Biome controller: CONTROLLER_DEFAULT
[13:26:50] [main/INFO]: [CHAT] §e    Feature: FEATURE_LAKES
[13:26:50] [main/INFO]: [CHAT] §e    Feature: FEATURE_CANYONS
[13:26:50] [main/INFO]: [CHAT] §e    Sun brightness: 1.0
[13:26:50] [main/INFO]: [CHAT] §e    Star brightness: 1.0
[13:26:50] [main/INFO]: [CHAT] §e    Sky color: 1.0, 1.0, 1.0
[13:26:50] [main/INFO]: [CHAT] §e    Fog color: 1.0, 1.0, 1.0
```
I then deleted and re-created it once, and got the same dimension back. I then did it again, and got a different one. This time, I spawned underground below a flat frozen forest. Info:
```
[13:36:56] [main/INFO]: [CHAT] §e    Base seed: -9201105177961642346
[13:36:56] [main/INFO]: [CHAT] §e    World version: 1
[13:36:56] [main/INFO]: [CHAT] §e    Terrain: TERRAIN_SOLID
[13:36:56] [main/INFO]: [CHAT] §e        Base block: Stone
[13:36:56] [main/INFO]: [CHAT] §e        Canyon block: Oak Wood
[13:36:56] [main/INFO]: [CHAT] §e        Base fluid: Air
[13:36:56] [main/INFO]: [CHAT] §e    Biome controller: CONTROLLER_COLD
[13:36:56] [main/INFO]: [CHAT] §e    Feature: FEATURE_RAVINES
[13:36:56] [main/INFO]: [CHAT] §e    Feature: FEATURE_CANYONS
[13:36:56] [main/INFO]: [CHAT] §e    Sun brightness: 1.0
[13:36:56] [main/INFO]: [CHAT] §e    Star brightness: 1.0
[13:36:56] [main/INFO]: [CHAT] §e    Sky color: 1.0, 1.0, 1.0
[13:36:56] [main/INFO]: [CHAT] §e    Fog color: 1.0, 1.0, 1.0
[13:36:56] [main/INFO]: [CHAT] §e    Sky type: SKY_STARS2
```
Ever since then, I've only ever gotten the second one. I've never been able to get back to the first one. Shouldn't a world with the same seed and the same set of dimlets always produce the same dimension (assuming `randomizeSeed` is false, which it is)?

**Comment by McJty** [comment]
Only if you fully specify all dimlets because if you don't specify dimlets these are randomized and that randomization isn't using the seed

**Comment by josephcsible** [comment]
What are those ones randomized off of then? Shouldn't I be getting a different dimension every time then, since I'm not specifying any dimlets at all?

**Comment by McJty** [comment]
hmm would have to review the code again. Has been a while. I may actually be mistaken

**Comment by josephcsible** [comment]
Hmm, something's really weird. I deleted the whole run/ folder and tried it again. This time, I got two dimensions that are very similar, but not exactly the same (and both completely different than what I got last time):
![2017-12-27_14 21 07](https://user-images.githubusercontent.com/16061266/34391036-7f5d6078-eb11-11e7-9ef3-922be1c6dd87.png)
```
[14:21:20] [main/INFO]: [CHAT] §e    Base seed: -9201105177961642346
[14:21:20] [main/INFO]: [CHAT] §e    World version: 1
[14:21:20] [main/INFO]: [CHAT] §e    Terrain: TERRAIN_LOW_CAVERN
[14:21:20] [main/INFO]: [CHAT] §e        Base block: Stone
[14:21:20] [main/INFO]: [CHAT] §e        Scattered Orb blocks: Oak Wood
[14:21:20] [main/INFO]: [CHAT] §e        Scattered Orb blocks: Mossy Stone Bricks
[14:21:20] [main/INFO]: [CHAT] §e        Scattered Orb blocks: Acacia Wood Planks
[14:21:20] [main/INFO]: [CHAT] §e        Base fluid: Air
[14:21:20] [main/INFO]: [CHAT] §e    Biome controller: CONTROLLER_DEFAULT
[14:21:20] [main/INFO]: [CHAT] §e    Feature: FEATURE_LAKES
[14:21:20] [main/INFO]: [CHAT] §e    Feature: FEATURE_SCATTEREDORBS
[14:21:20] [main/INFO]: [CHAT] §e    Structure: STRUCTURE_JUNGLETEMPLE
[14:21:20] [main/INFO]: [CHAT] §e    Sun brightness: 1.0
[14:21:20] [main/INFO]: [CHAT] §e    Star brightness: 1.0
[14:21:20] [main/INFO]: [CHAT] §e    Sky color: 1.0, 1.0, 1.0
[14:21:20] [main/INFO]: [CHAT] §e    Fog color: 1.0, 1.0, 1.0
[14:21:20] [main/INFO]: [CHAT] §e    Celestial angle: 0.2
```
![2017-12-27_14 21 43](https://user-images.githubusercontent.com/16061266/34391037-811bf1d6-eb11-11e7-8f93-25dccf823dce.png)
```
[14:21:45] [main/INFO]: [CHAT] §e    Base seed: -9201105177961642346
[14:21:45] [main/INFO]: [CHAT] §e    World version: 1
[14:21:45] [main/INFO]: [CHAT] §e    Terrain: TERRAIN_LOW_CAVERN
[14:21:45] [main/INFO]: [CHAT] §e        Base block: Stone
[14:21:45] [main/INFO]: [CHAT] §e        Scattered Orb blocks: Oak Wood
[14:21:45] [main/INFO]: [CHAT] §e        Scattered Orb blocks: Mossy Stone Bricks
[14:21:45] [main/INFO]: [CHAT] §e        Scattered Orb blocks: Acacia Wood Planks
[14:21:45] [main/INFO]: [CHAT] §e        Base fluid: Air
[14:21:45] [main/INFO]: [CHAT] §e    Biome controller: CONTROLLER_DEFAULT
[14:21:45] [main/INFO]: [CHAT] §e    Feature: FEATURE_LAKES
[14:21:45] [main/INFO]: [CHAT] §e    Feature: FEATURE_SCATTEREDORBS
[14:21:45] [main/INFO]: [CHAT] §e    Structure: STRUCTURE_JUNGLETEMPLE
[14:21:45] [main/INFO]: [CHAT] §e    Sun brightness: 1.0
[14:21:45] [main/INFO]: [CHAT] §e    Star brightness: 1.0
[14:21:45] [main/INFO]: [CHAT] §e    Sky color: 1.0, 1.0, 1.0
[14:21:45] [main/INFO]: [CHAT] §e    Fog color: 1.0, 1.0, 1.0
[14:21:45] [main/INFO]: [CHAT] §e    Celestial angle: 0.2
```

**Comment by josephcsible** [comment]
Okay, I've made some headway. Here's what I've found:
1. Clear the run directory, start Minecraft, create a new world with -9201105177961642346 seed, create the dimension -> dimension with spikes of grass and flowers and trees and red clouds
2. Clear the run directory, start Minecraft, close Minecraft without creating a world, start Minecraft again, create a new world with -9201105177961642346 seed, create the dimension -> mostly flat stone dimension (completely dark once, bright but with slowness another time)
3. Clear the run directory, start Minecraft, create a new world with -9201105177961642346 seed, close and reopen Minecraft, load the saved world, create the dimension -> spawn deep underground in a flat snowy dimension with a galaxy skybox

Can you see if you reproduce these exact cases?

**Comment by josephcsible** [comment]
Okay, so I figured this out. Here's what's going on:
1. The documentation for HashMap says "This class makes no guarantees as to the order of the map; in particular, it does not guarantee that the order will remain constant over time."
2. Even though inserting the same elements in the same order usually results in them being iterated over consistently, they aren't for us, because DimletKey#hashCode() calls DimletType#hashCode(), which is just Object#hashCode(). This depends on JVM internals, which explains the weird behavior I saw.
3. In DimletRandomizer#setupWeightedRandomList(), we iterate through our Maps and add the contents to Lists. This "freezes" the strange ordering.
4. In RarityRandomSelector#select(), we use the seed to choose a random number, then use that random number to index into the list. Since the order is weird and inconsistent, so are the dimension's properties.
